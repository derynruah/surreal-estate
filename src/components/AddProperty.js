/* eslint-disable no-console */
import React, { useState } from "react";
// eslint-disable-next-line import/no-extraneous-dependencies
import axios from "axios";
import "../styles/add-property.css";
import Alert from "./Alert";

const AddProperty = () => {
  const initialState = {
    fields: {
      title: "",
      city: "Manchester",
      type: "Flat",
      bedrooms: "",
      bathrooms: "",
      price: "",
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);

  const [alert, setAlert] = useState(initialState.alert);

  const handleAddProperty = (e) => {
    e.preventDefault();

    setAlert({ message: "", isSuccess: false });

    axios
      .post("https://localhost:4000/v1/api/PropertyListing", fields)
      .then(() => setAlert({ message: "Property Added", isSuccess: true }))
      .catch(() =>
        setAlert({
          message: "Server error. Please try again later.",
          isSuccess: false,
        })
      );
  };

  const handleFieldChange = (e) => {
    setFields({ ...fields, [e.target.name]: e.target.value });
  };

  return (
    <div className="add-property">
      <form onSubmit={handleAddProperty}>
        <Alert message={alert.message} success={alert.isSuccess} />
        <div className="form-fields">
          <label htmlFor="title">
            Property Description
            <input
              id="title"
              name="title"
              value={fields.title}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="city">
            City
            <select
              id="city"
              name="city"
              value={fields.city}
              onChange={handleFieldChange}
            >
              <option value="Manchester">Manchester</option>
              <option value="Leeds">Leeds</option>
              <option value="Sheffield">Sheffield</option>
              <option value="Liverpool">Liverpool</option>
            </select>
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="type">
            Type
            <select
              id="type"
              name="type"
              value={fields.type}
              onChange={handleFieldChange}
            >
              <option value="Flat">Flat</option>
              <option value="Detached">Detached</option>
              <option value="Semi-Detached">Semi-Detached</option>
              <option value="Terraced">Terraced</option>
              <option value="End of Terrace">End of Terrace</option>
              <option value="Cottage">Cottage</option>
              <option value="Bungalow">Bungalow</option>
            </select>
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="bedrooms">
            Bedrooms
            <input
              type="number"
              min="0"
              id="bedrooms"
              name="bedrooms"
              value={fields.bedrooms}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="bathrooms">
            Bathrooms
            <input
              type="number"
              min="0"
              id="bathrooms"
              name="bathrooms"
              value={fields.bathrooms}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="price">
            Price
            <input
              type="number"
              min="0"
              step="0.01"
              id="price"
              name="price"
              value={fields.price}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <div className="form-fields">
          <label htmlFor="email">
            Email
            <input
              type="email"
              id="email"
              name="email"
              value={fields.email}
              onChange={handleFieldChange}
            />
          </label>
        </div>
        <button type="submit">Add</button>
      </form>
    </div>
  );
};

export default AddProperty;
