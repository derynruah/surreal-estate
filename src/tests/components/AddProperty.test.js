import React from "react";
import { render, screen } from "@testing-library/react";
import AddProperty from "../../components/AddProperty.js";

describe("AddProperty", () => {
  test("renders correctly", () => {
    const { asFragment } = render(<AddProperty />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders title feild", () => {
    render(<AddProperty />);
    const title = screen.getByLabelText(/property description/i);

    expect(title).toHaveAttribute("name", "title");
    expect(title).toHaveDisplayValue("");
  });

  test("renders city field", () => {
    render(<AddProperty />);
    const city = screen.getByLabelText(/city/i);

    expect(city).toHaveAttribute("name", "city");
    expect(city).toHaveDisplayValue("Manchester");
  });

  test("renders type field", () => {
    render(<AddProperty />);
    const type = screen.getByLabelText(/type/i);

    expect(type).toHaveAttribute("name", "type");
    expect(type).toHaveDisplayValue("Flat");
  });

  test("renders bedrooms field", () => {
    render(<AddProperty />);
    const bedrooms = screen.getByLabelText(/bedrooms/i);

    expect(bedrooms).toHaveAttribute("name", "bedrooms");
    expect(bedrooms).toHaveDisplayValue("");
    expect(bedrooms).toHaveAttribute("min", "0");
  });

  test("renders bathrooms field", () => {
    render(<AddProperty />);
    const bathrooms = screen.getByLabelText(/bathrooms/i);

    expect(bathrooms).toHaveAttribute("name", "bathrooms");
    expect(bathrooms).toHaveDisplayValue("");
    expect(bathrooms).toHaveAttribute("min", "0");
  });

  test("renders price field", () => {
    render(<AddProperty />);
    const price = screen.getByLabelText(/price/i);

    expect(price).toHaveAttribute("name", "price");
    expect(price).toHaveDisplayValue("");
    expect(price).toHaveAttribute("min", "0");
    expect(price).toHaveAttribute("step", "0.01");
  });

  test("renders email field", () => {
    render(<AddProperty />);
    const email = screen.getByLabelText(/email/i);

    expect(email).toHaveAttribute("name", "email");
    expect(email).toHaveDisplayValue("");
    expect(email).toHaveAttribute("type", "email");
  });
});
