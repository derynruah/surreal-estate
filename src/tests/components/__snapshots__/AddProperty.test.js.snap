// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <form>
      <div
        class="form-fields"
      >
        <label
          for="title"
        >
          Property Description
          <input
            id="title"
            name="title"
            value=""
          />
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="city"
        >
          City
          <select
            id="city"
            name="city"
          >
            <option
              value="Manchester"
            >
              Manchester
            </option>
            <option
              value="Leeds"
            >
              Leeds
            </option>
            <option
              value="Sheffield"
            >
              Sheffield
            </option>
            <option
              value="Liverpool"
            >
              Liverpool
            </option>
          </select>
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="type"
        >
          Type
          <select
            id="type"
            name="type"
          >
            <option
              value="Flat"
            >
              Flat
            </option>
            <option
              value="Detached"
            >
              Detached
            </option>
            <option
              value="Semi-Detached"
            >
              Semi-Detached
            </option>
            <option
              value="Terraced"
            >
              Terraced
            </option>
            <option
              value="End of Terrace"
            >
              End of Terrace
            </option>
            <option
              value="Cottage"
            >
              Cottage
            </option>
            <option
              value="Bungalow"
            >
              Bungalow
            </option>
          </select>
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="bedrooms"
        >
          Bedrooms
          <input
            id="bedrooms"
            min="0"
            name="bedrooms"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="bathrooms"
        >
          Bathrooms
          <input
            id="bathrooms"
            min="0"
            name="bathrooms"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="price"
        >
          Price
          <input
            id="price"
            min="0"
            name="price"
            step="0.01"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="form-fields"
      >
        <label
          for="email"
        >
          Email
          <input
            id="email"
            name="email"
            type="email"
            value=""
          />
        </label>
      </div>
      <button
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
